CoreScriptsRoot/ServerCoreScripts/ClientChat/ChatMain.lua

```diff
@@ -0,0 +1,763 @@
+		if (channel == "All") then
+			ChatWindow:SwitchCurrentChannel(channel)
+		end
+		
+		if (welcomeMessage ~= "") then
+			local messageObject = MessageLabelCreator:CreateWelcomeMessageLabel(welcomeMessage)
+			channelObj:AddMessageLabelToLog(messageObject)
+		end
+
+		DoFadeInFromNewInformation()
+	end	
+	
+end)
+
+EventFolder.OnChannelLeft.OnClientEvent:connect(function(channel)
+	ChatWindow:RemoveChannel(channel)
+
+	DoFadeInFromNewInformation()
+end)
+
+EventFolder.OnMuted.OnClientEvent:connect(function(channel)
+	--// Do something eventually maybe?
+	--// This used to take away the chat bar in channels the player was muted in.
+	--// We found out this behavior was inconvenient for doing chat commands though. 
+end)
+
+EventFolder.OnUnmuted.OnClientEvent:connect(function(channel)
+	--// Same as above.
+end)
+
+EventFolder.OnSpeakerExtraDataUpdated.OnClientEvent:connect(function(speakerName, data)
+	local speaker = SpeakerDatabase:GetSpeaker(speakerName)
+	if (not speaker) then
+		speaker = SpeakerDatabase:AddSpeaker(speakerName)
+	end
+	
+	for i, v in pairs(data) do
+		speaker[i] = v
+	end
+end)
+
+
+EventFolder.OnMainChannelSet.OnClientEvent:connect(function(channel)
+	if (ChatWindow:GetChannel(channel)) then
+		ChatWindow:SwitchCurrentChannel(channel)
+	end
+end)
+
+
+local reparentingLock = false
+local function connectGuiParent(GuiParent)
+	local DestroyGuardFrame = Instance.new("Frame")
+	DestroyGuardFrame.Name = "DestroyGuardFrame"
+	DestroyGuardFrame.BackgroundTransparency = 1
+	DestroyGuardFrame.Size = UDim2.new(1, 0, 1, 0)
+
+	for i, v in pairs(GuiParent:GetChildren()) do
+		v.Parent = DestroyGuardFrame
+	end
+	DestroyGuardFrame.Parent = GuiParent
+
+end
+
+connectGuiParent(GuiParent)
+
+GuiParent.Changed:connect(function(prop)
+	if (prop == "Parent" and not reparentingLock) then
+		reparentingLock = true
+
+		local children = GuiParent.DestroyGuardFrame:GetChildren()
+		for i, v in pairs(children) do
+			v.Parent = nil
+		end
+
+		LocalPlayer.CharacterAdded:wait()
+		GuiParent.Parent = PlayerGui
+
+		for i, v in pairs(children) do
+			v.Parent = GuiParent
+		end
+
+		connectGuiParent(GuiParent)
+
+		reparentingLock = false
+	end
+end)
+
+--// Always on top behavior that relies on parenting order of ScreenGuis
+--// This would end up really bad if something else tried to do the exact same thing however.
+PlayerGui.ChildAdded:connect(function(child)
+	if (child ~= GuiParent and not reparentingLock) then
+		reparentingLock = true
+
+		GuiParent.Parent = nil
+		RunService.RenderStepped:wait()
+		GuiParent.Parent = PlayerGui
+
+		reparentingLock = false
+	end
+end)
+
+
+EventFolder.GetInitDataRequest:FireServer()
+
+
+
+
+
+
+
+
+
+
+--////////////////////////////////////////////////////////////////////////////////////////////
+--///////////// Code to talk to topbar and maintain set/get core backwards compatibility stuff
+--////////////////////////////////////////////////////////////////////////////////////////////
+local Util = {}
```

Comment:
this is the same util signal I saw above, please move this into one implementation. What is missing from the other signal util that is not possible to use here?